sentinelhub/sh_utils.py

```diff
@@ -1,40 +1,102 @@
 """
 Module implementing some utility functions not suitable for other utility modules
 """
+from abc import ABC, abstractmethod
 from urllib.parse import urlencode
 
-from sentinelhub.config import SHConfig
-from sentinelhub.download import SentinelHubDownloadClient
 from sentinelhub.exceptions import MissingDataInRequestException
 
 
-def iter_pages(service_url, client=None, config=None, exception_message=None, **params):
-    """ Iterates over pages of items retrieved from Sentinel-Hub service
+class FeatureIterator(ABC):
+    """ An implementation of a base feature iteration class
+
+    Main functionalities:
+    - The iterator will load only as many features as needed at any moment
+    - It will keep downloaded features in memory so that iterating over it again will not have to download the same
+    features again.
     """
+    def __init__(self, client, url, params=None):
+        """
+        :param client: An instance of a download client object
+        :type client: DownloadClient
+        :param url: An URL where requests will be made
+        :type url: str
+        :param params: Parameters to be sent with each request
+        :type params: dict or None
+        """
+        self.client = client
+        self.url = url
+        self.params = params or {}
+
+        self.index = 0
+        self.features = []
+        self.next = None
+        self.finished = False
+
+    def __iter__(self):
+        """ Method called at the beginning of a new iteration
+
+        :return: It returns the iterator class itself
+        :rtype: FeatureIterator
+        """
+        self.index = 0
+        return self
+
+    def __next__(self):
+        """ Method called to provide the next feature in iteration
+
+        :return: the next feature
+        """
+        while self.index >= len(self.features) and not self.finished:
+            new_features = self._fetch_features()
+            self.features.extend(new_features)
```

Comment:
I would suggest using `collections.deque` for the `features`. Appending entries to `collections.deque` (O(1)) should be better then extending a list (memory reallocation when allocated mem is exhausted), particularly for when there is a large number of features ...

Reply:
The time complexity for extending a list and extending a `deque` is the same. You can check that [here](https://wiki.python.org/moin/TimeComplexity). They both have dynamic memory allocation, probably similar to Vectors in Java and C++.

Reply:
Yes, amortised extensions have the same complexity. Expanding a list, when list would outgrown its current memory allocation, consists of copying the whole array to a new (contiguous) block of memory. If original list is already large, this might take quite some time. `dequeue.extend` does not work with contiguous memory allocations, which means that there is no need for any memory reallocations. 

On the other hand, we probably won't be downloading the whole internet with these, so until we notice a performance hit, let's keep features as `list`.